src/cli/expand-patterns.js

```diff
@@ -0,0 +1,234 @@
+"use strict";
+
+const path = require("path");
+const fs = require("fs");
+const globby = require("globby");
+const fastGlob = require("fast-glob");
+const flat = require("lodash/flatten");
+
+/** @typedef {import('./util').Context} Context */
+
+/**
+ * @param {Context} context
+ */
+function* expandPatterns(context) {
+  const cwd = process.cwd();
+  const seen = Object.create(null);
+  let noResults = true;
+
+  for (const pathOrError of expandPatternsInternal(context)) {
+    noResults = false;
+    if (typeof pathOrError !== "string") {
+      yield pathOrError;
+      continue;
+    }
+
+    const relativePath = path.relative(cwd, pathOrError);
+
+    // filter out duplicates
+    if (relativePath in seen) {
+      continue;
+    }
+
+    seen[relativePath] = true;
+    yield relativePath;
+  }
+
+  if (noResults) {
+    // If there was no files and no other errors, let's yield a general error.
+    yield {
+      error: `No matching files. Patterns: ${context.filePatterns.join(" ")}`
+    };
+  }
+}
+
+// TODO: use `fast-glob` directly, without `globby`
+const baseGlobbyOptions = {
+  dot: true,
+  expandDirectories: false,
+  absolute: true
+};
+
+/**
+ * @param {Context} context
+ */
+function* expandPatternsInternal(context) {
+  // Ignores files in version control systems directories and `node_modules`
+  const ignoredDirectories = {
+    ".git": true,
+    ".svn": true,
+    ".hg": true,
+    node_modules: context.argv["with-node-modules"] !== true
+  };
+
+  const globbyOptions = {
+    ...baseGlobbyOptions,
+    ignore: Object.keys(ignoredDirectories)
+      .filter(dir => ignoredDirectories[dir])
+      .map(dir => "**/" + dir)
+  };
+
+  let supportedFilesGlob;
+  const cwd = process.cwd();
+
+  /**
+   * @type {Array<{
+   *  type: 'file' | 'dir' | 'glob';
+   *  absolutePath?: string;
+   *  glob?: string;
+   *  input: string;
+   * }>}
+   */
+  const entries = [];
+
+  for (const pattern of context.filePatterns) {
+    const absolutePath = path.resolve(cwd, pattern);
+
+    if (containsIgnoredPathSegment(absolutePath, cwd, ignoredDirectories)) {
+      continue;
+    }
+
+    const stat = statSafeSync(absolutePath);
+    if (stat) {
+      if (stat.isFile()) {
+        entries.push({ type: "file", absolutePath, input: pattern });
+      } else if (stat.isDirectory()) {
+        entries.push({
+          type: "dir",
+          absolutePath,
+          glob: escapePathForGlob(pattern) + "/" + getSupportedFilesGlob(),
+          input: pattern
+        });
+      }
+    } else if (pattern[0] === "!") {
+      // convert negative patterns to `ignore` entries
+      globbyOptions.ignore.push(fixWindowsSlashes(pattern.slice(1)));
+    } else {
+      entries.push({
+        type: "glob",
+        glob: fixWindowsSlashes(pattern),
+        input: pattern
+      });
+    }
+  }
+
+  for (const { type, absolutePath, glob, input } of entries) {
```

Comment:
Do we need try dir first? For dir we ignore NoParserError, but for glob we print the error. What do we do with `prettier dir "dir/**/*"` and `prettier "dir/**/*" dir`? 

Still in hospital, can't play with it, just posting my concerts.

Reply:
I haven't implemented parser inference for dirs yet. But these are good test cases.

Reply:
> What do we do with `prettier dir "dir/**/*"` and `prettier "dir/**/*" dir`?

I think this is okay:
```console
$ ../../../bin/prettier.js dir1 'dir1/**/*' -l
dir1/a1.js
dir1/b1.js
dir1/nested1/an1.css
dir1/nested1/an1.js
dir1/nested1/bn1.js
dir1/a1.foo[error] No parser could be inferred for file: dir1/a1.foo
$ ../../../bin/prettier.js 'dir1/**/*' dir1 -l
dir1/a1.foo[error] No parser could be inferred for file: dir1/a1.foo
dir1/a1.js
dir1/b1.js
dir1/nested1/an1.css
dir1/nested1/an1.js
dir1/nested1/bn1.js
```